pkg/util/error.go

```diff
@@ -43,15 +43,23 @@ func IsNotFound(err error) bool {
 	return false
 }
 
-func IsConnectionRefusedOrUnauthorized(err error) bool {
+func IsConnectionRefusedOrUnauthorizedOrAskingForCredentials(err error) bool {
 	if err == nil {
 		return false
 	}
 
 	msg := strings.ToLower(err.Error())
-	if strings.Contains(msg, "connection refused") || strings.Contains(msg, "unauthorized") {
+	if strings.Contains(msg, "connection refused") || strings.Contains(msg, "unauthorized") || strings.Contains(msg, "the server has asked for the client to provide credentials") {
```

Comment:
use human readable error messages to decide specific error cases in general is bad practice. Because the dependent upstream library may change the context of those errors, it's not reliable, instead we should always try to find the typed error in the depend library so that if the definition of the error changed, we are aware in the code base.

The reason why we accept `connection refused` is because this error is coming from go `net/http` package, if you check the [source code](https://cs.opensource.google/go/go/+/master:src/net/http/socks_bundle.go;l=209), it's very low level error code, in the socket layer, basically the author will not change it.

However, the `"unauthorized"` is already a code smell, unauthorized is a standard HTTP status code - 401, you could find in the `net/http/status.go` [here](https://cs.opensource.google/go/go/+/master:src/net/http/status.go;l=37) have all status code with const value defined, instead of using string to compare, we should use this const value. Because this error response are coming from API server, so the related library to define the response is apimachinery, so a much more reliable way is to case this error to [`StatusError`](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/errors/errors.go#L35), then we can extract the status reason to tell if this error is unauthorized or not.

Then let's see this newly introduced message, actually, if we did the Unauthorized error correctly, we don't need to introduce this comparing, if you look at where is this message defined, it's from apimachinery [here](https://github.com/kubernetes/apimachinery/blob/master/pkg/api/errors/errors.go#L454), and the HTTP status of this message is exactly `StatusReasonUnauthorized`

So please refactor this code, use HTTP status to check Unauthorized error but not continuing to introduce this message comparing bad practices.

@c-pius @nesmabadr, if you need more clarity for this refactoring, feel free to reach me.